File: vhosts/blog/nuldoc-src/generators/post_list.ts
import { renderToDOM } from "../jsx/render.ts";
import PostListPage from "../pages/PostListPage.tsx";
import { Config } from "../config.ts";
import { Page } from "../page.ts";
import { PostPage } from "./post.ts";

export type PostListPage = Page;

export async function generatePostListPages(
  posts: PostPage[],
  config: Config,
): Promise<PostListPage[]> {
  const postsPerPage = config.blog.postsPerPage;
  const totalPages = Math.ceil(posts.length / postsPerPage);
  const pages: PostListPage[] = [];

  for (let pageIndex = 0; pageIndex < totalPages; pageIndex++) {
    const pagePosts = posts.slice(
      pageIndex * postsPerPage,
      (pageIndex + 1) * postsPerPage,
    );

    const page = await generatePostListPage(
      pagePosts,
      config,
      pageIndex + 1,
      totalPages,
    );

    pages.push(page);
  }

  return pages;
}

async function generatePostListPage(
  posts: PostPage[],
  config: Config,
  currentPage: number,
  totalPages: number,
): Promise<PostListPage> {
  const html = await renderToDOM(
    PostListPage(
      posts,
      config,
      currentPage,
      totalPages,
    ),
  );

  const destFilePath = currentPage === 1
    ? "/posts/index.html"
    : `/posts/${currentPage}/index.html`;

  const href = currentPage === 1 ? "/posts/" : `/posts/${currentPage}/`;

  return {
    root: html,
    renderer: "html",
    destFilePath,
    href,
  };
}

File: services/nuldoc/lib/nuldoc/generators/post_list.rb
module Nuldoc
  module Generators
    class PostList
      def initialize(posts, config)
        @posts = posts
        @config = config
      end

      def generate
        posts_per_page = @config.sites.blog.posts_per_page
        total_pages = (@posts.length.to_f / posts_per_page).ceil
        pages = []

        (0...total_pages).each do |page_index|
          page_posts = @posts[page_index * posts_per_page, posts_per_page]
          current_page = page_index + 1

          html = Pages::PostListPage.render(
            posts: page_posts,
            config: @config,
            current_page: current_page,
            total_pages: total_pages
          )

          dest_file_path = current_page == 1 ? '/posts/index.html' : "/posts/#{current_page}/index.html"
          href = current_page == 1 ? '/posts/' : "/posts/#{current_page}/"

          pages.push(Page.new(
                       root: html,
                       renderer: :html,
                       site: 'blog',
                       dest_file_path: dest_file_path,
                       href: href
                     ))
        end

        pages
      end
    end
  end
end
